// @llamaindex/groq/dist/index.cjs
Object.defineProperty(exports, '__esModule', { value: true });

var env = require('@llamaindex/env');
var openai = require('@llamaindex/openai');
var GroqSDK = require('groq-sdk');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var GroqSDK__default = /*#__PURE__*/_interopDefault(GroqSDK);

// Models that support tool/function calling
const TOOL_CALLING_MODELS = [
    "gemma2-9b-it",
    "llama-3.1-8b-instant",
    "llama-3.3-70b-versatile",
    "deepseek-r1-distill-llama-70b",
    "qwen-qwq-32b",
    "meta-llama/llama-4-scout-17b-16e-instruct",
    "meta-llama/llama-4-maverick-17b-128e-instruct"
];
class Groq extends openai.OpenAI {
    constructor(init){
        const { apiKey = env.getEnv("GROQ_API_KEY"), additionalSessionOptions = {}, model = "llama-3.3-70b-versatile", ...rest } = init ?? {};
        super({
            apiKey,
            additionalSessionOptions,
            model,
            ...rest
        });
        this.lazySession = async ()=>new GroqSDK__default.default({
                apiKey,
                ...init?.additionalSessionOptions
            });
    }
    get supportToolCall() {
        return TOOL_CALLING_MODELS.includes(this.model);
    }
}
/**
 * Convenience function to create a new Groq instance.
 * @param init - Optional initialization parameters for the Groq instance.
 * @returns A new Groq instance.
 */ const groq = (init)=>new Groq(init);

exports.Groq = Groq;
exports.groq = groq;
